easybuild/easyblocks/generic/cargopythonbundle.py

```diff
@@ -0,0 +1,57 @@
+##
+# Copyright 2018-2023 Ghent University
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for installing python bundles with some packages using Rust
+
+@author: Mikael Oehman (Chalmers University of Technology)
+"""
+
+from easybuild.easyblocks.generic.cargo import Cargo
+from easybuild.easyblocks.generic.pythonbundle import PythonBundle
+
+
+class CargoPythonBundle(PythonBundle, Cargo):  # PythonBundle must come first to take precedence
+    """
+    Builds just like PythonBundle with setup for Rust and crates from Cargo easyblock
+
+    The cargo init step will set up the environment variables for rustc and vendor sources
+    but all the build steps are triggered like normal.
+    """
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Define extra easyconfig parameters specific to Cargo"""
+        extra_vars = PythonBundle.extra_options(extra_vars)
+        extra_vars = Cargo.extra_options(extra_vars)  # not all extra options here will used here
+
+        return extra_vars
+
+    def __init__(self, *args, **kwargs):
+        """Constructor for CargoPythonBundle easyblock."""
+        super(CargoPythonBundle, self).__init__(*args, check_for_sources=False, **kwargs)
```

Comment:
@Micket Hmm, this doesn't work because `Cargo` doesn't have support for `check_for_sources`...

```
  File "/Users/kehoste/work/easybuild-easyblocks/easybuild/easyblocks/generic/cargopythonbundle.py", line 53, in __init__
    super(CargoPythonBundle, self).__init__(*args, check_for_sources=False, **kwargs)
  File "/Users/kehoste/work/easybuild-easyblocks/easybuild/easyblocks/generic/pythonbundle.py", line 59, in __init__
    super(PythonBundle, self).__init__(*args, **kwargs)
  File "/Users/kehoste/work/easybuild-easyblocks/easybuild/easyblocks/generic/bundle.py", line 75, in __init__
    super(Bundle, self).__init__(*args, **kwargs)
  File "/Users/kehoste/work/easybuild-easyblocks/easybuild/easyblocks/generic/cargo.py", line 94, in __init__
    super(Cargo, self).__init__(*args, **kwargs)
  File "/Users/kehoste/work/easybuild-framework/easybuild/framework/extensioneasyblock.py", line 96, in __init__
    EasyBlock.__init__(self, *args, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'check_for_sources'
```

Reply:
I forgot where i left this off, i tried a bunch of different variants in python2 and python3, struggling to find a way to worked for both. 

I've been swamped with other work.  Feel free to push whatever you want on this branch or take over it completely if you have a better idea to solve this issue.

Reply:
So, having looked at this again, trying a few things, using kwargs here is just completely impossible, again thanks for the MRO in python. So. yeah. No way to do this apart from say, setting a variable in `self` and checking if it's defined in Bundle.  Yuck.

```python3
    def __init__(self, *args, **kwargs):
        """Constructor for CargoPythonBundle easyblock."""
        self.check_for_sources = False
        super(CargoPythonBundle, self).__init__(*args, **kwargs)
```
and in Bundle
```
    def __init__(self, *args, **kwargs):
        """Constructor for Bundle easyblock"""
        self.check_for_sources = False
        if not hasattr(self, 'check_for_sources'):
            self.check_for_sources = True
        
        if self.check_for_sources:
            ...
```


Alternatively, one can make Bundle less draconian and just allow sources, at least if it's not a plain Bundle class.
```
if type(self) is Bundle:
    # check for sources only here
```